### PR TITLE
feat(android): use cameraX

### LINKED ISSUE #13525: Android: CameraX
Add cameraX support to the Titanium Android SDK

### Solution

Related PR https://github.com/tidev/titanium_mobile/pull/12395

### Platforms

Android

### PR DESCRIPTION
**Pull request by m1ga**
**JIRA:** https://jira-archive.titaniumsdk.com/AC-6676
Fixes https://github.com/tidev/titanium_mobile/issues/13525

**Optional Description:**

Will use CameraX on Android when using a camera with an overlay

```javascript
const win = Ti.UI.createWindow({layout: "vertical", backgroundColor: "#fff"});

const buttons = [
	"open camera (old camera)",
	"open camera",
	"open camera (no autoHide)",
	"open camera (flash on)",
	"open camera (flash auto)",
	"open camera (video)",
	"open camera (video, low bit rate)",
	"open camera (video, low resolution)",
	"open camera (video, max duration)",
	"open camera (small output image)",
];
const results = Ti.UI.createView({height: Ti.UI.SIZE, width: Ti.UI.FILL, layout: "horizontal"});
const img = Ti.UI.createImageView({width: "45%", height: Ti.UI.SIZE, autorotate: true});
const videoPlayer = Titanium.Media.createVideoPlayer({height: 200, width: "45%", mediaControlStyle: Titanium.Media.VIDEO_CONTROL_DEFAULT, scalingMode: Titanium.Media.VIDEO_SCALING_RESIZE_ASPECT});
var isRecording = false;
var videoMaximumDuration = 0;
var isBack = true;

results.add([img, videoPlayer])

for (var i = 0; i < buttons.length; ++i) {
	var btn = Ti.UI.createButton({title: buttons[i], idx: i + 1, height: 40, borderRadius:0, width: 300});
	btn.addEventListener("click", openCamera);
	win.add(btn);
}
win.add(results);

var cameraSizes = Ti.Media.cameraOutputSizes;
if (cameraSizes != null) {
	for (var i=0; i<cameraSizes.length;++i){
		console.log(cameraSizes[i].cameraType);
		console.log(cameraSizes[i].values);
	}
}


function openCamera(e) {
	var sourceId = e.source.idx;
	var isVideo = (sourceId == "6" || sourceId == "7" || sourceId == "8" || sourceId == "9");
	var overlay = Ti.UI.createView({height: Ti.UI.FILL, height: Ti.UI.FILL});
	var isTorch = isFlash = isZoom = false;
	var btnText = (isVideo) ? "Record video" : "Take picture";
	var useCameraX = (sourceId != "1");
	var btn = Ti.UI.createButton({title: btnText, bottom: 10});
	var btn2 = Ti.UI.createButton({title: "Switch cam", bottom: 50});
	var btn3 = Ti.UI.createButton({title: "torch", bottom: 90, left: 10});
	var btn4 = Ti.UI.createButton({title: "flash (off)", bottom: 90, right: 10});
	var btn5 = Ti.UI.createButton({title: "zoom", bottom: 90});
	overlay.add([btn, btn2, btn3, btn4,btn5]);

	btn.addEventListener("click", function() {
		if (isVideo) {
			if (!isRecording) {
				Ti.Media.startVideoCapture()
				isRecording = true;
				btn.title = "Stop";
			} else {
				Ti.Media.stopVideoCapture()
				isRecording = false;
				btn.title = btnText;
			}
		} else {
			Ti.Media.takePicture()
		}
	})
	btn2.addEventListener("click", function() {
		if (isBack) {
			Ti.Media.switchCamera(Titanium.Media.CAMERA_FRONT)
		} else {
			Ti.Media.switchCamera(Titanium.Media.CAMERA_REAR)
		}
		isBack = !isBack;
	})
	btn3.addEventListener("click", function() {
		isTorch = !isTorch;
		Ti.Media.torch = isTorch;
	})
	btn4.addEventListener("click", function() {
		if (isFlash) {
			Ti.Media.cameraFlashMode = Titanium.Media.CAMERA_FLASH_OFF
			btn4.title = "flash (off)";
		} else {
			Ti.Media.cameraFlashMode = Titanium.Media.CAMERA_FLASH_ON
			btn4.title = "flash (on)";
		}
		isFlash = !isFlash;
	})
	btn5.addEventListener("click", function() {
		if (isZoom) {
			Ti.Media.zoomLevel = 1;
		} else {
			Ti.Media.zoomLevel = Ti.Media.maxZoomLevel - 0.5;
		}
		isZoom = !isZoom;
	})

	var flashMode = Titanium.Media.CAMERA_FLASH_OFF;
	if (sourceId == "4") {
		isFlash = true;
		flashMode = Titanium.Media.CAMERA_FLASH_ON;
		btn4.title = "flash (on)";
	} else if (sourceId == "5") {
		flashMode = Titanium.Media.CAMERA_FLASH_AUTO;
	}

	var autohide = true;
	if (sourceId == "3") {
		autohide = false;
	}

	var mediaType = Titanium.Media.MEDIA_TYPE_IMAGE;
	if (sourceId == "6") {
		mediaType = Titanium.Media.MEDIA_TYPE_VIDEO;
	}

	var bitRate = -1;
	var frameRate = -1;
	if (sourceId == "7") {
		mediaType = Titanium.Media.MEDIA_TYPE_VIDEO;
		bitRate = 120;
		frameRate = 2 * 1024 * 1024;
	}

	var maxResolution = null;
	if (sourceId == "8") {
		mediaType = Titanium.Media.MEDIA_TYPE_VIDEO;
		maxResolution = {
			width: 300,
			height: 200
		}
	}

	if (sourceId == "9") {
		mediaType = Titanium.Media.MEDIA_TYPE_VIDEO;
		videoMaximumDuration = 2000;
	}

	var imgWidth = -1;
	var imgHeight = -1;
	if (sourceId == "10") {
		// lower the target image resolution
		imgWidth = 240;
		imgHeight = 320;
	}

	btn2.visible = btn3.visible = btn4.visible = btn5.visible = useCameraX;

	Ti.Media.requestCameraPermissions(function(e) {
		if (e.success) {
			Ti.Media.showCamera({
				autohide: autohide,
				useCameraX: useCameraX,
				cameraFlashMode: flashMode,
				overlay: overlay,
				frameRate: frameRate,
				bitRate: bitRate,
				mediaTypes: [mediaType],
				saveToPhotoGallery: true,
				maxResolution: maxResolution,
				videoMaximumDuration: videoMaximumDuration,
				targetImageWidth: imgWidth,
				targetImageHeight: imgHeight,
				success: function(e) {

					if (e.mediaType == "public.movie") {
						var f = Ti.Filesystem.getFile(Ti.Filesystem.applicationDataDirectory, "tmp.mp4");
						f.write(e.media);
						videoPlayer.url = f.nativePath;
						videoPlayer.play();
					} else {
						img.image = e.media;
					}
					console.log("type", e.mediaType);
					console.log("size", e.media.size);
					console.log("width", e.media.width, "height", e.media.height);
				}
			});
		}
	});
}

win.open();
```

**Features:**
- [x] open camera
- [x] close camera
- [x] add overlay
- [x] switch front/back
- [x] take picture
- [x] video support
- [x] mediatype
- [x] new image size parameters: `targetImageWidth` and `targetImageHeight`
- [x] new `Ti.Media.cameraOutputSizes` to get the camera output sizes
- [x] video quality
- [x] video max duration
- [x] flash
- [x] torch
- [x] pinch to zoom (photo & video)
- [x] `minZoomLevel`, `maxZoomLevel` and `zoomLevel` to set the zoom in code
- [x] check file destination / saveToGallery
- [x] clean up code
- [x] `open` callback with preview size return values
- [x] `aspectRatio` with `Ti.Media.ASPECT_RATIO_16_9` and `Ti.Media.ASPECT_RATIO_4_3`
- [x]  `scalingMode` with `Ti.Media.IMAGE_SCALING_ASPECT_FIT` and `Ti.Media.IMAGE_SCALING_ASPECT_FILL`
- [x] `verticalAlign` with `Ti.Media.VERTICAL_ALIGN_BOTTOM`, `Ti.Media.VERTICAL_ALIGN_TOP`, `Ti.Media.VERTICAL_ALIGN_CENTER`
- [x] add `useCameraX` parameter

**Tested:**
* Pixel 4, Android 13
* Pixel 7, Android 13
* Samsung A20, Android 11

```js
Ti.Media.showCamera({
	success: onCameraSuccess,
	overlay: view_overlay,
	aspectRatio: Ti.Media.ASPECT_RATIO_16_9,
	scalingMode: Ti.Media.IMAGE_SCALING_ASPECT_FIT,
	open: function(e) {
		console.log(e);
	}
});
```

<b>16:9 image with ASPECT_FILL</b>
image

<b>16:9 image with ASPECT_FIT</b>
Screenshot_20230319-161953

### CODE REVIEW

**Comment by hansemannn** [comment]
@m1ga Big one. Will you likely finish it or should be close it? I'd be happy to take it if it's stable. But maybe through a module first? Like we did with Ti.WKWebView.

**Comment by m1ga** [comment]
I think we should keep this open. Everything was working fine already just couldn't write the file to the gallery but I guess that is due to the new permissions. 
The main purpose was to remove the old stuff from the SDK 😄

**Comment by m1ga** [comment]
great, latest camerax version changed a lot in the video part. Have to rewrite that part. But it looks like you can record without audio permission now :+1:

**Comment by m1ga** [comment]
I can now:
* record a video to a file and to the media gallery (and still get the file in Ti)
* capture an image to file and to the media gallery (and still get the file in Ti)
* added a `recording` callback to get some infos (duration, size) during a video recording

now testing and some error feedback

**Comment by m1ga** [comment]
Ready to test :rocket: 
I've tested it on a Pixel 4 (Android 12) and a Samsung A5 (Android 8)

File: android/titanium/build.gradle
Comment (hansemannn) [inline android/titanium/build.gradle:284]: Nice! Finally in RC!

Comment (m1ga) [reply android/titanium/build.gradle:284]: yes! And only some bug fixes so hopefully it's final soon :crossed_fingers:

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:132]: Move to:
```java
if (successCallback == null) {
    return;
}
```
to prevent some deep nesting.

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:214]: This looks like duplicate logic from above at [L111](https://github.com/tidev/titanium_mobile/pull/12395/files#diff-52633e11fe636615f24f4c067fc7d0dadd18e950519f41d4b159905616aaee87R135)

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:294]: Add implementation or remove / add note?

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:324]: Handle exception

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:318]: No need for example brackets

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:353]: Add comment?

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:515]: Combine if-statements

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:517]: The US-english version is preferred (`canceled`).

File: android/titanium/build.gradle
Comment (hansemannn) [inline android/titanium/build.gradle:290]: 2.1.4 is available: https://developer.android.com/jetpack/androidx/releases/constraintlayout

File: android/titanium/build.gradle
Comment (hansemannn) [inline android/titanium/build.gradle:284]: Note: 1.2.0 beta 2 is available, just let's do that in a later release.

**Review by hansemannn — CHANGES_REQUESTED** [review]
Some feedback! But looking good, I think we can merge this very soon!

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (m1ga) [reply android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:214]: kind of. Just with `MediaStore.Video`

File: android/titanium/build.gradle
Comment (m1ga) [reply android/titanium/build.gradle:284]: :+1: yes, they've changed a lot in the 1.1.0 betas so we'll wait for the release in order not to change it multiple times :smile:

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (m1ga) [reply android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:515]: Think that won't work because the `finish()` is called in the else part called even if `cancelCallback` is null

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (m1ga) [reply android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:111]: I think I can move it down to the `imageCapture.takePicture` line. Need to test that

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [reply android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:515]: Oh, I missed the `finish` call!

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (m1ga) [reply android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:111]: try-catch was just needed for the `getTempFile` part. I've put it there and return if it would fail. Looks better :+1:

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (m1ga) [reply android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:297]: Missed that, will implement it!

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (m1ga) [reply android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:327]: It's from https://github.com/tidev/titanium_mobile/blob/81d3f4a1bfa69a5367b8cf05af40b608b62ebe3b/android/titanium/src/java/org/appcelerator/titanium/io/TiContentFile.java#L427-L439
it will return `null` when the try-catch fails

**Comment by m1ga** [comment]
added `Ti.Media.torch = true/false` to switch on the light when the camera is open. That will light up the whole time and not like the flash only when you take the image

**Comment by m1ga** [comment]
Added output image parameters:

```js
var cameraSizes = Ti.Media.cameraOutputSizes;
for (var i=0; i<cameraSizes.length;++i){
	console.log(cameraSizes[i].cameraType);
	console.log(cameraSizes[i].values);
}
```
to get the potential sizes. 

And you can set them with:
```js
targetImageWidth: 800,
targetImageHeight: 600,
```

**Comment by m1ga** [comment]
some new additions that makes working with overlays a lot easier:

* `open` callback that contains the size of the preview (e.g. if you use 4:3 it will be wider than your screen)
* `aspectRatio` with two constants `Ti.Media.ASPECT_RATIO_16_9` and  `Ti.Media.ASPECT_RATIO_4_3` 
* `scalingMode`  with `Ti.Media.IMAGE_SCALING_ASPECT_FIT` and `Ti.Media.IMAGE_SCALING_ASPECT_FILL` 

imageScreenshot_20230319-161953

having the preview only be FIT (full width) makes it easier to work with overlays:

**Comment by m1ga** [comment]
* `verticalAlign` with `Ti.Media.VERTICAL_ALIGN_BOTTOM` (top, bottom, center)


Screenshot_20230319-163636 Screenshot_20230319-163647

**Comment by m1ga** [comment]
added the old files back. The new cameraX part is used when you set `useCameraX: true` opening the camera.
That way we can merge this quicker as the default is still the old way and you can opt-in to the new way

**Comment by hansemannn** [comment]
Let's take this in and give it some testing during the next 1-2 releases. If it runs smooth, we can make it default after that 🙏

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:230]: Log error

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:281]: The `isRecording` property is set to `true` both times and can therefore be after the condition.

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:287]: Are inline if-conditions recommended in our style guide? In JS, it's mostly disallowed these days

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:326]: Do we use this one in our CLI as well? Just to be sure.

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (hansemannn) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:568]: Any more strong references, e.g. `callbackContext` and `overlayProxy`? It would be good to run the profiler for any memory leaks.

**Review by hansemannn — CHANGES_REQUESTED** [review]
Some more feedback - mostly questions to be sure! One general thing: As the code is pretty separated now, wouldn't it make sense to wrap it up inside a module to be able to quickly iterate on it, then include it to the SDK once stable? We did the same with ti.wkwebview back then.

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (m1ga) [reply android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:333]: I took that from https://github.com/tidev/titanium_mobile/blob/3362dd2154714c817185096708c7a4b7e4b5bf6f/android/modules/media/src/java/ti/modules/titanium/media/MediaModule.java#L859 to create a custom filename that is without any special characters

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java
Comment (m1ga) [reply android/modules/media/src/java/ti/modules/titanium/media/TiCameraXActivity.java:575]: Good point. I'll check that. The current TiCameraActivity doesn't do more so if that works fine it will be  the same in TiCameraXActivity.

**Comment by m1ga** [comment]
Thanks for the feedback! Already adjusted most of it!

>  to wrap it up inside a module to be able to quickly iterate on it, then include it to the SDK once stable? We did the same with ti.wkwebview back then.

I did so much work here already that I don't want to rip it out and create a module. I think more people test it if we just say "enable useCameraX" and don't have to add a module. It would be different too (no `Ti.Media...` namespace) so people would need to change code after it's merged into the SDK.

My idea would be: 
have people **test it in the RC/beta phase to see that it is working as a replacement** (using the current features, just with the flag enabled) and then we know it is working without a difference.
 
If the **new features have to be changed it will be in a minor bugfix release** as they are "new features" that won't break existing apps. There wasn't any demand for those features so they aren't time critical (and they are Android only).

**Comment by hansemannn** [comment]
Totally fine! Let's merge it and await the feedback!

**Review by hansemannn — APPROVED** [review]
Amazing work, really excited to use this in production apps soon!

**Comment by m1ga** [comment]
awesome! Thanks!

**Pull request merged**
